cpp/src/arrow/type.h

```diff
@@ -1256,6 +1257,10 @@ class ARROW_EXPORT DictionaryType : public FixedWidthType {
 
   bool ordered() const { return ordered_; }
 
+  /// \brief Determine whether dictionary arrays may be compared without unification
+  /// (smaller dictionaries are prefixes of larger dictionaries)
+  static Result<bool> ComparableWithoutUnification(std::vector<const Array*> arrays);
```

Comment:
It may be a nitpick but I would prefer to put this function into `arrow/array/dictionary_util.h` or `arrow/array/array_dict.h` (since splitting up `array.h` is going to happen eventually)

I might prefer a two-argument version of this. 

```
namespace dictionary {

bool CanCompareIndices(const DictionaryArray& left, const DictionaryArray& right);

}
```

Reply:
```
DictionaryArray::CanCompareIndices
```

would be OK too

Reply:
I'll refactor to that